CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+
+**Operator Overloading**
+
+Logical AND ("&&"), logical OR ("||") and the comma (",") operators
+must never be overloaded.
+
+**Derived Classes**
+
+A derived class must never redefine a non-virtual function.
+
+**Downcasting**
+
+Downcasting must always be given justification.  Generally, a downcast represents 
+a problematic design.
+
+**Return Types**
+
+Returning a non-const reference or pointer to non-public member data from a
+public class member function requires justification.
+
+**Function/Operator Overloading**
+
+Function or operator overloading must only be used when performing the same
+logical function.
+
+**Inheritance**
+
+Behavioral type consistency must be maintained between publicly derived
+classes and base classes. That is, public inheritance must implement an
+"is-a" relationship and subtypes (publicly derived classes) must be
+substitutable.
+
+Use of multiple inheritance or virtual inheritance (that is, using the virtual
+keyword in the inheritance specification for a class to make one or more of its
+parent classes be virtual base classes) requires justification.
+
+**Complexity**
+
+Code functions must not be overly complex.
+
+**Dead Code**
+
+Production software must not include any dead code.
+
+Dead code is: *Executable object code (or data) that is part of the production
+software which, as a result of a design error cannot be executed or used in an
+operational configuration of the target computer environment and is not
+traceable to a system or software requirement.*
+
+**Unused Code**
+
+Unused source code must be prevented from inclusion in production software
+by means of preprocessor directives, and clearly documented as unused source
+code by means of code comments. Code statements must not be commented out in
+production source code unless they are clearly part of the source code
+documentation (e.g. examples).
+
+Unused code is: *Compile time excluded non-executable source code. That is,
+source code that is discarded by the compilation process and does not result in
+production executable object code (or data).*
+
+**Goto**
+
+The `goto` keyword should not be used.
+
+**Variable Initialization**
+
+Variables with automatic storage duration must be initialized before use.
+
+**Global Variables**
+
+Use of variables with external linkage ("global variables") should not be used.
+
+**Recursion**
+
+Code that implements recursive algorithms must be supported by a justification
+showing that the execution time and memory requirements are bounded and
+compatible with the target hardware and software architecture.
+
+**Unused Return Values**
+
+The return value of a non-void function must be explicitly cast to (void)
+using a C-style cast if the value is not used by the calling function.
+
+[I believe the purpose here is to show that anyone ignoring the return value 
+of a function is doing so intentionally, but I'm considering pulling it anyway
+because it's a little weird.  Thoughts?]
```

Comment:
I have never seen that before, but I think it is meant to encourage using error codes:

if (!file.open()){
  //handle error here
}

Reply:
You're correct.  There are more specific examples in a different part of the standard, though I think I cut that part because it was talking about alternatives to exception handling.

```
[[nodiscard]]
int Foo() { return 3; }

int main()
{
    (void)Foo();
}
```

With C++17, the [[dodiscard]] tag will tell the compiler to generate a warning if the return value is ignored, *unless* the return value is *explicitly discarded*.  I think.

Reply:
Attribute tags are part of the C++17 standard, so we can use them if and only if we're compiling on C++17.

Reply:
Would it be possible to run them as part of a testing suite? As in, tests are compiled with C++17 features(where they make sense, like this one), and the actual binary on the robot is C++14?